docs/rfcs/027-pageserver-migration.md

```diff
@@ -0,0 +1,454 @@
+  S3 content during the migration that will never be read.
+
+The user expectations for availability are:
+
+- For planned maintenance, there should be zero availability
+  gap
+- For unplanned changes (e.g. node failures), there should be
+  minimal availability gap.
+
+## Non Goals (if relevant)
+
+- We do not aim to have the pageservers fail over if the
+  control plane is unavailable.
+- On unplanned migrations (node failures), we do not aim to prevent a small, bounded window of
+  read unavailability of very recent LSNs (because postgres
+  page cache usually contains such pages, we do not expect
+  them to be read frequently from the pageserver).
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+Pageserver, control plane
+
+## Terminology
+
+- **Attachment**: a tenant is _attached_ to a pageserver if it has
+  been issued a generation number, and is running an instance of
+  the `Tenant` type, ingesting the WAL, and available to serve
+  page reads.
+- **Location**: locations are a superset of attachments. A location
+  is a combination of a tenant and a pageserver. We may _attach_ at a _location_.
+
+- **Secondary location**: a location which is not currently attached.
+- **Warm secondary location**: a location which is not currently attached, but is endeavoring to maintain a warm local cache of layers. We avoid calling this a _warm standby_ to avoid confusion with similar postgres features.
+
+## Implementation (high level)
+
+### Secondary locations
+
+To enable faster migrations, we will identify at least one _secondary location_
+for each tenant. This secondary location will keep a warm cache of layers
+for the tenant, so that if it is later attached, it can catch up with the
+latest LSN quickly.
+
+The control plane is responsible for selecting secondary locations, and
+calling into pageservers to instruct them as to which tenants they are
+to act as a secondary location for.
+
+The attached pageserver for a tenant will publish a [layer heatmap](#layer-heatmap)
+to advise secondaries of which layers should be downloaded.
+
+### Location modes
+
+Currently, we consider a tenant to be in one of two states on a pageserver:
+
+- Attached: active `Tenant` object, and layers on local disk
+- Detached: no layers on local disk, no runtime state.
+
+We will extend this with finer-grained modes, whose purpose will become
+clear in later sections:
+
+- **AttachedSingle**: equivalent the existing attached state.
+- **AttachedMulti**: like AttachedSingle, holds an up to date generation, but
+  does not do deletions.
+- **AttachedStale**: like AttachedSingle, holds a stale generation,
+  do not do any remote storage operations.
+- **Secondary**: keep local state on disk, periodically update from S3.
+- **Detached**: equivalent to existing detached state.
+
+To control these finer grained states, a new `tenant/<tenant id>/configure`
+pageserver API will be introduced.
```

Comment:
We already have the tenant config API.

Reply:
Updated this to define a separate `tenant/<id>/location/config` endpoint (and in the detail section below)